exploration/dataflow-composability.md

```diff
@@ -0,0 +1,792 @@
+and formatted results. This proposal just defines wrappers
+for those types that implementations must use in order to
+make custom functions as flexible as possible.
+
+## Use-Cases
+
+Use cases from [issue 515](https://github.com/unicode-org/message-format-wg/issues/515):
+
+The following code fragment
+invokes the `:number` formatter on the literal `1`,
+binds the result to `$a`, and then invokes `:number` on the
+value bound to `$a`.
+
+If the value of `$a` does not allow for inspecting the previous options
+passed to the first call to `:number`,
+then the `$b` would format as `1.000`.
+
+### Example 1.1
+```
+.local $a = {1 :number minIntegerDigits=3} // formats as 001.
+.local $b = {$a :number minFractionDigits=3} // formats as 001.000
+// min integer digits are preserved from the previous call.
+```
+
+In other words: the user likely expects this code to be equivalent to:
+
+### Example 1.2
+```
+.local $b = {1 :number minIntegerDigits=3 minFractionDigits=3}
+```
+
+But without `:number` being able to access the previously passed options,
+the two fragments won't be equivalent.
+This requires `:number` to return a value that encodes
+the options that were passed in, the value that was passed in,
+and the formatted result;
+not just the formatted result.
+
+This example is an instance of the basic motivator for this proposal:
+allowing data that flows out of a function call
+to flow back into another function call
+with all of its metadata (e.g. options) preserved.
+
+### Example 1.3
+```
+.input {$item :noun case=accusative count=1}
+.local $colorMatchingGrammaticalNumberGenderCase = {$color :adjective accord=$item}
+```
+
+The `:adjective` function is a hypothetical custom formatter.
+If the value of its `accord` option is a string, it's hard for `:adjective`
+to use the value of `accord` to inflect the value of `$color` appropriately
+given the value of `$item`.
+We want to pass not the formatted result of `{$item :noun case=accusative count=1}`
+into `:adjective`, but rather, a structure that encodes that formatted result,
+along with the resolved value of `$item` and the names and values of the options
+previously passed to `:noun`: `case` and `count`.
+
+### Example 1.4
+
+```
+.local $foo = {$arg :func}
+```
+
+Here, `$arg` is treated as an input variable.
+Suppose that internally, an implementation wants to pre-define
+some formatting options on all input variables
+(or all input variables whose values have a particular type).
+It would be helpful if functions could accept
+a single argument that wraps the value of `$arg`
+along with these predefined options,
+separate from the options that are specific to the function.
+
+## Requirements
+
+- Define the structure passed in as an argument to a custom formatting function.
+- Define the structure that a custom formatting function should return.
+- Maintain the options passed into the callee as a _separate_ argument to the
+  formatter, to avoid confusion. (See Example 4 below.)
```

Comment:
I had somewhat different thoughts. 

```
.input {$v1 :f1 o1=a...}
.local $v2 {$v1 :f2 o1=a o2=b}

```
- f1 == f2. For the **same** function (or aliases with different options), do a merge on the option map (with the later ones winning). This could be extended to alias functions (the same under the covers, with different option settings).
- **f1 ≠ f2**. For **different** functions, it gets tricky. I think we need to require it to be specified in the registry.

Example:
.input {$v1 :number o1:a}
.local $v2 {$v1 :date o1=a o2=b}

The registry could specify that :date can handle a resolved :number expression in the following way.
- The value of the :date expression is the number converted to a double value and interpreted as the number of seconds since 1970-01-01T00:00:00Z.
- The number operands are discarded except for o3=x, which is mapped to o9 in the following way: ...

Reply:
@macchiati noted:

> I had somewhat different thoughts.

My thoughts have evolved here also. I think it isn't the function or isn't _just_ the function that determines whether an option is inherited. I think options should be specified as inheritable (or not).

Consider:

```
.input {$d :datetime year=numeric month=long day=numeric}
.local $t = {$d :datetime hour=numeric minute=numeric}
{{The event is on {$d} at {$t}.}}
```

These are the _same_ input variables and the _same_ function. You don't want to inherit skeleton/fields.

Now consider:

```
.input {$d :datetime dateStyle=long timeZone=$userZone numberingSystem=Latn}
.local $t = {$d :datetime timeStyle=long}
{{The event is on {$d} at {$t}.}}
```

Again, you don't want to inherit the styles. But you *do* want to inherit the coerced time zone and the numbering system (because it makes more sense to inherit it).

> The registry could specify that :date can handle a resolved :number expression in the following way.

The registry can already specify that. And _implementations_ can already do that, as we allow implementation defined types. The `:number` options won't matter to `:date` (etc.) and not be inherited.

Note that `:datetime`, `:date`, and `:time` are not the same function (f1 ≠ f2) but probably inherit/share some fields when "composed" as in the examples above (replace `:datetime` with `:date` and `:time` for example). The current registry doesn't do this because the potentially shared fields were pushed down into the RGI bucket for 45.

Reply:
I think it depends on the example. In the following, I do want to inherit.
```
.input {$d :datetime year=numeric month=numeric day=numeric}
.local $d2 = {$d :datetime month=long}
{{The event is on {$d} at {$d2}.}}
```
What I think would be cleaner would be to have options like suppress=date, suppress=time. These cause a host of options to be set to none. Then the following would be very clear as to what is to happen, _especially for the poor reader who doesn't want to have to consult the registry for each attribute to find out whether it inherits or not_.

```
.input {$d :datetime year=numeric month=long day=numeric}
.local $t = {$d :datetime suppress=date hour=numeric minute=numeric}
{{The event is on {$d} at {$t}.}}

```

Reply:
I don't have time right now to comment in depth, but I really want to discourage, again, the use of the word "inherit" here, since options are not related to object-oriented inheritance. I think the right thing to do instead is to define data types that standard functions return, and exhaustively specify the required and optional fields in those data types; if an option doesn't appear there, then it can't be used in composing functions. (And then, custom function writers get a way to define these types for themselves.)

There is no escaping that we're talking about describing dynamic data in a static way, so this is like defining a "data model" for runtime values to make it practical to understand and describe how functions interoperate. In such a model, options are passed as components of a data structure; not inherited.

Reply:
> Does that print "Your name is ADDIS..." or "Your name is Addis..." or what?

I think that must depend on the `:text-transform` and `:truncate` implementations, and how their authors intend for the values to compose. I think we must allow for each function to make choices when determining their resolved values about the "value" and "options" they present.

Something very similar is achievable even with our most basic built-in functions, and without options:
```
.local $x = {4.2e1 :number}
{{{$x :string}}}
```

If `:number` presents its input directly as its value, that should format as "4.2e1", but if it presents a numeric value, then that should format as "42". I do not think enforcing the first on a spec level makes much sense.

Similarly, a "text-transform" could effectively consume its `operation=uppercase` option when constructing its value, or consider it a string formatting option. So either of these are conceivable in the resolved value:
1. Value `'Addison'`, options `{ operation: 'uppercase' }`
2. Value `'ADDISON'`, options `{}`

As we've discussed in a variety of contexts, not all options are really the same, so I'd say that enabling this freedom is the right choice.